File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Ribbon/UiRibbonButton.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { Window } from "../Control/UiWindow";
import { ButtonType } from "../UiCommon";
import { IVisual } from "../UiVisual";

export interface IRibbonButton extends IControl {
    new (window: Window, key?: string | number): IRibbonButton;

    SetText(s: string): RibbonButton;
    GetText(): string;

    SetSmall(b: boolean): RibbonButton;
    GetSmall(): boolean;

    SetCheck(b: boolean): RibbonButton;
    GetCheck(): boolean;

    SetButtonType(n: ButtonType): RibbonButton;
    GetButtonType(): ButtonType;

    SetDrop(c: IControl): IControl;
    GetDrop(): IControl;

    SetBigVisual(v: IVisual): IVisual;
    GetBigVisual(): IVisual;

    SetSmallVisual(v: IVisual): IVisual;
    GetSmallVisual(): IVisual;

    OnClick(fn: (sender: RibbonButton) => void): RibbonButton;
    OnDrop(fn: (sender: RibbonButton) => void): RibbonButton;
}

export class RibbonButton extends (AveLib.UiRibbonButton as IRibbonButton) {
    private m_Drop: IControl;
    private m_BigVisual: IVisual;
    private m_SmallVisual: IVisual;

    SetDrop(c: IControl) {
        this.m_Drop = c;
        return super.SetDrop(c);
    }

    SetBigVisual(v: IVisual) {
        this.m_BigVisual = v;
        return super.SetBigVisual(v);
    }

    SetSmallVisual(v: IVisual) {
        this.m_SmallVisual = v;
        return super.SetSmallVisual(v);
    }
}

File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiRibbonButton.cpp
#include "StdAfx.h"
#include "UiRibbonButton.h"

#define ThisMethod($x) &UiRibbonButton::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiRibbonButton::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiRibbonButton> c_obj;
	}

	void UiRibbonButton::DefineControl()
	{
		AutoAddMethod( SetText );
		AutoAddMethod( GetText, WrapObjectGeneric );
		AutoAddMethod( SetSmall );
		AutoAddMethod( GetSmall, WrapObjectGeneric );
		AutoAddMethod( SetCheck );
		AutoAddMethod( GetCheck, WrapObjectGeneric );
		AutoAddMethod( SetButtonType );
		AutoAddMethod( GetButtonType, WrapObjectGeneric );
		AutoAddMethod( SetDrop );
		AutoAddMethod( GetDrop, WrapObjectGeneric );
		AutoAddMethod( SetBigVisual );
		AutoAddMethod( GetBigVisual, WrapObjectGeneric );
		AutoAddMethod( SetSmallVisual );
		AutoAddMethod( GetSmallVisual, WrapObjectGeneric );
		AutoAddMethod( OnClick, WrapObjectGeneric );
		AutoAddMethod( OnDrop, WrapObjectGeneric );
	}

	U1 UiRibbonButton::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	void UiRibbonButton::__OnClick( Ui::IRibbonButton & sender )
	{
		m_OnClick( this );
	}

	void UiRibbonButton::__OnDrop( Ui::IRibbonButton & sender )
	{
		m_OnDrop( this );
	}

	UiRibbonButton* UiRibbonButton::OnClick /**/( Callback_t&& fn ) { SetEventCallback<Ui::IRibbonButton::OnClick /**/>( m_OnClick /**/, std::move( fn ), MakeThisFunc( __OnClick /**/ ) ); return this; }
	UiRibbonButton* UiRibbonButton::OnDrop  /**/( Callback_t&& fn ) { SetEventCallback<Ui::IRibbonButton::OnDrop  /**/>( m_OnDrop  /**/, std::move( fn ), MakeThisFunc( __OnDrop  /**/ ) ); return this; }
}
